### ISSUE TITLE
SQL state code returned on duplicate foreign key is inconsistent

### ISSUE THREAD

**Issue by jmf-tls** [cockroachdb/cockroach#42858]
**Problem description**
SQL state code returned on duplicate foreign key is inconsistent with PostgreSQL and among different CockroachDB versions.
This SQL state code is returned by the JDBC driver, when running the queries from a Java application (including, for example, a DB manager like DBeaver).

**To Reproduce**
With a CockroachDB (or PostgreSQL) instance running, execute:
```
-- Create 2 tables
CREATE TABLE TEST1 (COL1 SERIAL PRIMARY KEY, COL2 INT8);
CREATE TABLE TEST2 (COL1 SERIAL PRIMARY KEY, COL2 INT8);

-- Create a foreign key twice
ALTER TABLE TEST1 ADD CONSTRAINT XPTO FOREIGN KEY (COL2) REFERENCES TEST2 (COL1);
ALTER TABLE TEST1 ADD CONSTRAINT XPTO FOREIGN KEY (COL2) REFERENCES TEST2 (COL1);
```
The results are different depending on the DB used:  

| DB  | SQL state |
| ------------- | ------------- |
| CockroachDB v19.2.x  | 23503 - ERROR: column "col2" cannot be used by multiple foreign key constraints  |
| CockroachDB v19.1.x  | 42830 - ERROR: columns cannot be used by multiple foreign key constraints  |
| PostgreSQL v9 or v10 | 42710 - ERROR: constraint "xpto" for relation "test1" already exists  |

**Expected behavior**
The SQL state code should be consistent, at least among CockroachDB versions.
Furthermore, if the codes mean:
 -  42710 - ERRCODE_DUPLICATE_OBJECT - constraint xxxx already exists.
 -  23503 - ERRCODE_FOREIGN_KEY_VIOLATION - a column in a table has  a value that is not present in the column of other table in the foreign key relation
 - 42830 - ERRCODE_INVALID_FOREIGN_KEY - the types of columns in the relation are incompatible

It seems to me that the most adequate code is 42710, used by PostgreSQL.

**Comment by jmf-tls** [comment]
I suppose the correct error might have been caught in this func
https://github.com/cockroachdb/cockroach/blob/1c3fce73cd25fa69e5a8c05ce8e215d66b4f49e4/pkg/sql/sqlbase/table.go#L309
but the `info` is only being filled if `index.ID` corresponds to the primary key, or if `index.Unique`, which I believe Foreign Keys are neither?...

**Comment by jordanlewis** [comment]
@otan could you take a look, maybe with @lucy-zhang? thanks!

**Comment by otan** [comment]
Just before I dig further:

> The SQL state code should be consistent, at least among CockroachDB versions.

> It seems to me that the most adequate code is 42710, used by PostgreSQL.

I incline to agree with you here, but there is a possibility of 20.1 then having different behaviour to 19.2 (unless we backport it in a 19.2.2 release). any opinions on this?

**Comment by jmf-tls** [comment]
@otan yes, you've got this dilemma...
For human readable feedback, the ERROR message is fine as it can always point to the concrete problem regardless of the error code.
Personally, I prefer the error code to be consistent with PostgreSQL, however I checked MySQL and SQL server and each uses a different code when this happens. I'd say the most important is to keep consistency, by making sure the error code won't change anymore, either by keeping the 23503 used since 19.2.0, or by sticking with 42710 from 19.2.2 onwards.

But besides all this, can you check that code on `collectConstraintInfo` that I mentioned in the previous comment?
I'm not exactly an expert on Go, but it seems that the duplicate name of the PK should have been caught in that func.
The `if index.ID == desc.PrimaryIndex.ID {...} else index.Unique {...}` seems to prevent `info` from getting filled with data in lots of cases; if `info` had data, the error code 42710 would be returned from the following line:
https://github.com/cockroachdb/cockroach/blob/1c3fce73cd25fa69e5a8c05ce8e215d66b4f49e4/pkg/sql/sqlbase/table.go#L358
What do you think?

**Comment by otan** [comment]
Apologies for the delay in reply.

> The if index.ID == desc.PrimaryIndex.ID {...} else index.Unique {...} seems to prevent info from getting filled with data in lots of cases; if info had data, the error code 42710 would be returned from the following line:

The code you are linking appears to be populating information -- not the path where we attempt to insert new rows. I believe what we want to do is add the error check here -- https://sourcegraph.com/github.com/cockroachdb/cockroach@14ea66f0e1a64fd7ba617c532bafbe379c1094a5/-/blob/pkg/sql/create_table.go#L407.

This seems to be a genuine error in the grand scheme of things -- i.e.
```
root@127.0.0.1:57913/movr>
CREATE TABLE TEST1 (COL1 SERIAL PRIMARY KEY, COL2 INT8, COL3 INT8);
CREATE TABLE TEST2 (COL1 SERIAL PRIMARY KEY, COL2 INT8);

-- Create a foreign key twice
ALTER TABLE TEST1 ADD CONSTRAINT XPTO FOREIGN KEY (COL2) REFERENCES TEST2 (COL1);
ALTER TABLE TEST1 ADD CONSTRAINT XPTO FOREIGN KEY (COL3) REFERENCES TEST2 (COL1);
CREATE TABLE

...

ERROR: duplicate index name: "test1_auto_index_xpto"
```

We should add a check in `ResolveFK` to make sure we don't allow foreign keys with the same name twice. For context, this automatic index creation error will not trigger whenever we create a foreign key with the same name (#37255) and will probably allow the re-name to happen, and as such we cannot rely on the check for "same column" and "same index used" to catch this for us.

tl;dr I will submit a PR shortly to return the desired `42710` code with an extra check - but I think I will leave this behaviour to be consistent from 20.1 onwards.

**Comment by knz** [comment]
I would use code 23503 and also backport this change to 19.2.

**Comment by lucy-zhang** [comment]
@otan I'm not sure if we're talking about the same thing, but my guess for what's happening is that we check for duplicate index names (in `ValidateTableIndexes()`) _before_ we check for duplicate FK names (which I think happens in some caller of `ValidateTable()`, which in turn calls `CheckUniqueConstraints()`), which obviously produces unexpected results when we auto-generate indexes for FKs.

I have no idea why we don't just check for duplicate FK names in or near `ResolveFK()` itself. My best guess is that we auto-generate FK names somewhere else (if no name is explicitly provided), and that's why we leave uniqueness checking until the end, but that's just a guess. Maybe this is a relic of when FKs used to live on index descriptors and it was more annoying to perform the check.

Basically, adding a check like you're describing in `ResolveFK` doesn't seem unreasonable to me, but the code around detecting duplicate names seems illogical in general. Maybe someone else has some insight into this.

**Comment by otan** [comment]
> I have no idea why we don't just check for duplicate FK names in or near ResolveFK() itself. 
It seems to be checked in `Validate()`, which is commented out: https://github.com/cockroachdb/cockroach/blob/91bc9ec64f1321ca39fe9838fea584db455726ab/pkg/sql/alter_table.go#L288

But it wouldn't matter because, as you say, primary key checks come first!

>  My best guess is that we auto-generate FK names somewhere else (if no name is explicitly provided), and that's why we leave uniqueness checking until the end, but that's just a guess.
Yep, it seems you can't name a FK unless you make `ALTER TABLE` do it - so I think this is why we leave it to the end / at the `Validate` self.

> Basically, adding a check like you're describing in ResolveFK doesn't seem unreasonable to me, but the code around detecting duplicate names seems illogical in general. Maybe someone else has some insight into this.

I'm going to put this in `alter_table.go`, since I think names can only be generated with the `ALTER TABLE` syntax.
